### PR TITLE
scripts: update: improve log readability fix issues and add BLE library update ...

### PR DESCRIPTION
**Pull request by ABOSTM**
This PR improves python scripts for stm32 cube updates :

* scripts: serie_update.py and update_stm32_package : run black -l 80 
* scripts: update_stm32_package: add optional call to genllheaders 
* scripts: cube update: add option to force a specific version 
* scripts: cube update: Improve log readability and add debug option 
* scripts: update: detect any unclean repo and ask user about continuing 
* scripts: serie_update: fixes update conflict error 
* scripts: genllheaders: keep only ll header file name ending with .h 
* scripts: update_stm32_package: commit generated generic LL headers 
* scripts: update: add stm32wb ble library update

### CODE REVIEW

File: scripts/serie_update.py [hunk 1]
```diff
@@ -142,6 +149,24 @@ def os_cmd(self, cmd, cwd=None, shell=False):
...
         )
+
+    def rename_conf_template(self, path):
+        """renames hal_conf_template.h to hal_conf.h ...
+        Args:
+            path: path where to apply the files processing
+        """
+        hal_conf_template = [
+            f for f in path.iterdir() if "hal_conf_template.h" in f.name
+        ][0]
```
Comment (gmarull) [inline scripts/serie_update.py:161]: why iterate to obtain a known file name?

File: scripts/genllheaders/genllheaders.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def main(hal_path, output):
         # obtain series LL headers
         series_headers = entry / "drivers" / "include"
         for header in series_headers.iterdir():
-            m = re.match(r"stm32[a-z0-9]+_ll_([a-z0-9]+)\.h", header.name)
+            m = re.match(r"stm32[a-z0-9]+_ll_([a-z0-9]+)\.h$", header.name)
```
Comment (gmarull) [inline scripts/genllheaders/genllheaders.py:71]: `^` may also be added at the beginning

File: scripts/update_stm32_package.py [hunk 3]
```diff
@@ -195,4 +195,41 @@ def update_cubes():
     genllheaders.main(
         REPO_ROOT / "stm32cube", REPO_ROOT / "stm32cube" / "common_ll"
     )
+
+    # commit autogenerate generic LL HAL headers
+    if args.force:
+        zephyr_hal_stm32_path = (
+            Path(os.getenv("ZEPHYR_BASE")).absolute()
```
Comment (gmarull) [inline scripts/update_stm32_package.py:202]: see previous comment

**Review by gmarull — CHANGES_REQUESTED** [review]

File: scripts/genllheaders/genllheaders.py [hunk 4]
```diff
@@ -34,9 +34,7 @@
 REPO_ROOT = SCRIPT_DIR / ".." / ".."
 """Repository root (used for input/output default folders)."""
 
-LL_API_IGNORE = [
-    "usb"
-]
+LL_API_IGNORE = ["usb"]
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/genllheaders/genllheaders.py:37]: Ok, thanks. So I changed and run black with default 88 cols.

File: scripts/update_stm32_package.py [hunk 5]
```diff
@@ -112,3 +113,19 @@ def update_cubes():
...
+    REPO_ROOT = (
+        Path(os.getenv("ZEPHYR_BASE")).absolute()
+        / ".."
+        / "modules"
+        / "hal"
+        / "stm32"
+    )
+    genllheaders.main(
+        REPO_ROOT / "stm32cube", REPO_ROOT / "stm32cube" / "common_ll"
+    )
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/update_stm32_package.py]: Reworked

File: scripts/serie_update.py [hunk 6]
```diff
@@ -127,6 +115,33 @@ def __init__(
...
+        Args:
+            cmd: string command to execute.
+            cwd: directory where to run command
+            shell: boolean to enable command interpretation by the shell
+
+        Returns:
+            return the returncode of the command after execution.
+        """
+        if self.debug:
+            logging.debug(str(cmd) + "      cwd:" + str(cwd))
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: For no good reason :-)
Fixed

File: scripts/update_stm32_package.py [hunk 7]
```diff
@@ -131,6 +132,39 @@ def update_cubes():
         logging.info("Add option --noclean")
         args.noclean = True
 
+# prevent losing data in an unclean repo
+scripts_repo_path = (
+    Path(os.getenv("ZEPHYR_BASE")).absolute()
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/update_stm32_package.py]: Reworked

File: scripts/update_stm32_package.py [hunk 8]
```diff
@@ -131,6 +132,39 @@ def update_cubes():
...
+for status in git_status:
+    if not status.decode("utf-8").startswith("?") and not status.decode(
+        "utf-8"
+    ).startswith("!"):
+        logging.error(
+            "It seems that script repo "
+            + str(scripts_repo_path)
+            + "\n           is not clean:\n           git status -s:"
+            + status.decode("utf-8")
+        )
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/update_stm32_package.py]: Done across whole files

File: scripts/serie_update.py (hunk 1 above)
Comment (ABOSTM) [reply scripts/serie_update.py]: Inheritance of initial scripts. Reworked.

File: scripts/genllheaders/genllheaders.py (hunk 2 above)
Comment (ABOSTM) [reply scripts/genllheaders/genllheaders.py]: Done

File: scripts/serie_update.py [hunk 9]
```diff
@@ -14,6 +14,7 @@
 from pathlib import Path
 import logging
 from jinja2 import Environment, FileSystemLoader
+from packaging import version
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py:17]: Thanks @gmarull for your review.
I fixed all your remark, except this one which is not clear to me:
yes, it seems "packaging" is used for handling python modules version.
but I don't think it is exclusive, and it make the job properly for stm32cube version.
So do you see any restriction to use it for cube version ? is there a better alternative ?

**Comment by ABOSTM** [comment]
^^ @FRASTM

File: scripts/serie_update.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: Ok I get ride of packaging module.

File: scripts/serie_update.py [hunk 10]
```diff
@@ -191,9 +190,7 @@ def get_zephyr_current_version(self):
         else:
             self.clean_files()
             raise ValueError(
-                "Error: cannot find version "
-                + previous_version
-                + " in STM32Cube_repo"
+                "Error: cannot find version " + previous_version + " in STM32Cube_repo"
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: Done

File: scripts/serie_update.py [hunk 11]
```diff
@@ -283,10 +276,11 @@ def extract_source(self):
...
-                str(stm32cube_drivers_src_path) + "/*.* " +
-                str(temp_drivers_src_path)
+                "cp "
+                + "-r "
+                + str(stm32cube_drivers_src_path)
+                + "/*.* "
+                + str(temp_drivers_src_path)
             ),
             shell=True,
         )
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: I made it in an additional commit.

File: scripts/serie_update.py [hunk 12]
```diff
@@ -303,9 +297,7 @@ def build_from_current_cube_version(self):
 
         # build the zephyr module from the stm32cube
         self.extract_source()
-        logging.info(
-            "Building module from STM32Cube_repo " + self.current_version
-        )
+        logging.info("Building module from STM32Cube_repo " + self.current_version)
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: I used now f-string whenever possible

File: scripts/serie_update.py [hunk 13]
```diff
@@ -474,14 +466,10 @@ def update_cmakelist(self):
             source_files.sort()
 
             cmakelists_new.write(
-                "zephyr_library_sources(soc/system_"
-                + self.stm32_seriexx
-                + ".c)\n"
+                "zephyr_library_sources(soc/system_" + self.stm32_seriexx + ".c)\n"
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: Done

File: scripts/serie_update.py [hunk 14]
```diff
@@ -659,11 +644,7 @@ def apply_zephyr_patch(self):
...
                 "-r",
                 str(self.stm32cube_temp_serie),
-                str(
-                    self.zephyr_hal_stm32_path
-                    / "stm32cube"
-                    / self.stm32_seriexx
-                ),
+                str(self.zephyr_hal_stm32_path / "stm32cube" / self.stm32_seriexx),
             )
         )
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: Done in dedicated commit

File: scripts/serie_update.py [hunk 15]
```diff
@@ -474,14 +466,10 @@ def update_cmakelist(self):
...
-                "zephyr_library_sources(soc/system_"
-                + self.stm32_seriexx
-                + ".c)\n"
+                "zephyr_library_sources(soc/system_" + self.stm32_seriexx + ".c)\n"
             )
             cmakelists_new.write(
-                "zephyr_library_sources(drivers/src/"
-                + self.stm32_seriexx
-                + "_hal.c)\n"
+                "zephyr_library_sources(drivers/src/" + self.stm32_seriexx + "_hal.c)\n"
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: Done

File: scripts/update_stm32_package.py [hunk 16]
```diff
@@ -91,7 +91,11 @@ def update_cubes():
     raise Exception("ZEPHYR_BASE Not defined")
 
 if not args.noclean:
-    print("Do you want to clean downloaded repo (" + str(repo_path) + ") at the end of updates?")
+    print(
+        "Do you want to clean downloaded repo ("
+        + str(repo_path)
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/update_stm32_package.py]: Done

File: scripts/serie_update.py [hunk 17]
```diff
@@ -668,6 +617,17 @@ def apply_zephyr_patch(self):
             )
         )
 
+        # apply dos2unix to whole zephyr hal serie sub directory
+        for child in self.zephyr_module_serie_path.glob("**/*"):
+            if child.is_file:
+                self.os_cmd(("dos2unix", child), cwd=self.zephyr_module_serie_path)
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py:595]: Because I am working on windows (with git-bash), and at some point is show delta linked to CR/LF (I don't remember exactly but I think it depends on stm32 series repo).

File: scripts/ble_library.py [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+This script updates ble library
+"""
+
+import sys
+import subprocess
+from pathlib import Path
+import logging
+from jinja2 import Environment, FileSystemLoader
+
+include_dir = ["stm32wb/hci"]
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/ble_library.py:16]: Today there is only one include directory, but having a list allow flexibility and potentially several path in the future

File: scripts/ble_library.py [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+
+
+def copy_hci_files(src_repo_path, dest_lib_path):
+    """Copy sources files from Cube Firmware to zephyr"""
+    # remove existing *.c and *.h files
+    subprocess.check_call(
+        ("rm", "*.c", "*.h"),
+        shell=False,
+        cwd=Path(dest_lib_path / "stm32wb" / "hci"),
+    )
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/ble_library.py]: Done

File: scripts/ble_library.py [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+            # copy each file to destination
+            subprocess.check_call(
+                (
+                    "cp",
+                    file_path,
+                    Path(dest_lib_path / "stm32wb" / "hci" / file_name),
+                ),
+                shell=False,
+                cwd=dest_lib_path,
+            )
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/ble_library.py]: Done

File: scripts/ble_library.py [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+    # Create file from Jinja2 template
+    templates_dir = dest_lib_path / ".." / "scripts"
+    j2_env = Environment(
+        loader=FileSystemLoader(str(templates_dir)),
+        trim_blocks=True,
+        lstrip_blocks=True,
+    )
+    README_j2_template = j2_env.get_template("ble_README_template.txt")
+    README_path = dest_lib_path / "stm32wb" / "hci" / "README"
+    with open(README_path, "w") as README:
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/ble_library.py]: Done

File: scripts/ble_library.py [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+        )
+
+
+def update(src_repo_path, dest_lib_path, version, commit):
+    """Update ble library"""
+    logging.info(" ... Updating ble library ...")
+    copy_hci_files(src_repo_path, dest_lib_path)
+    apply_ble_patches(dest_lib_path)
+    create_ble_lib_README(dest_lib_path, version, commit)
+    create_ble_lib_CMakeList(dest_lib_path)
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/ble_library.py]: Done

File: scripts/ble_library.py [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+    "Middlewares/ST/STM32_WPAN/utilities/stm_list.h",
+    "Middlewares/ST/STM32_WPAN/utilities/utilities_common.h",
+    "Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_Hid/Core/Inc/app_common.h",
+    "Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_Hid/Core/Inc/app_conf.h",
+    "Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_Hid/Core/Inc/hw_if.h",
+    "Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_Hid/STM32_WPAN/Target/"
+    + "hw_ipcc.c",
+    "Projects/P-NUCLEO-WB55.Nucleo/Applications/BLE/BLE_Hid/STM32_WPAN/App/"
+    + "tl_dbg_conf.h",
+]
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/ble_library.py:42]: It is not really helpful is this case because most of the case I ake only part of a directory. 
It could be useful only for  "Middlewares/ST/STM32_WPAN/interface/patterns/ble_thread/shci/" .. but It contains only to file
So I keep as is, which is at least explicit on the copied files.

File: scripts/update_stm32_package.py [hunk 24]
```diff
@@ -132,6 +129,33 @@ def update_cubes():
...
+        )
+        logging.error("%s", "It is suggested to clean your repo first.")
+
+        print("Do you want to continue update ?")
+        res = input("(Enter y/n) ").lower()
+        while res not in ("y", "n"):
+            res = input("(Enter y/n) ").lower()
+        if res == "n":
+            sys.exit()
+        break
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/update_stm32_package.py:155]: Ooops, I missed this one. I will implement it.

[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/update_stm32_package.py:155]: This looks interesting, but finally, I wonder whether this is a good idea, because pygit2 is not part of zephyr requirements.
So adding this dependencies seems too much only for that (we are currently very few persons to update stm32 cube), especially knowing that I already have a working solution (even if more complex).

**Comment by ABOSTM** [comment]
@erwango @FRASTM can you please review ?

File: scripts/ble_README_template.txt [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+
+        * Move some files license from ST SLA0044 to BSD-3 Clause
+            Following files are distributed under ST License SLA0044.
+            For Zephyr they are made available under BSD-3 Clause.
+            - hw_if.h
+            - hw_ipcc.c
+            - app_common.h
+            - app_conf.h
+            - stm32_wpan_common.h
+            - ble_bufsize.h
```
Comment (erwango) [inline scripts/ble_README_template.txt:50]: MIssing tl_dbg_conf.h

File: lib/ble_zephyr.patch [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+index bf9ea29f..91dd8fef 100644
+--- a/lib/stm32wb/hci/ble_bufsize.h
++++ b/lib/stm32wb/hci/ble_bufsize.h
+@@ -8,9 +8,9 @@
+  * All rights reserved.</center></h2>
+  *
+- * This software component is licensed by ST under Ultimate Liberty license
+- * SLA0044, the "License"; You may not use this file except in compliance with
+- * the License. You may obtain a copy of the License at:
+- *                             www.st.com/SLA0044
```
Comment (erwango) [inline lib/ble_zephyr.patch:11]: Why is this file required ?
In usual cube packages updates, we're able to report fixes from one version to the next w/o any .patch file

File: scripts/ble_library.py [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+
+import sys
+from pathlib import Path
+import logging
+from jinja2 import Environment, FileSystemLoader
+
+include_dir = ["stm32wb/hci"]
+
+file_list = [
+    "Middlewares/ST/STM32_WPAN/ble/core/ble_bufsize.h",
```
Comment (erwango) [inline scripts/ble_library.py:19]: What are the modifications required to this file when a new patch is done to this library?
I don't think lib update automation should bring complexity to usual development flow..

**Review by erwango — CHANGES_REQUESTED** [review]
Looks good, but I feel the wb hci ble lib automation is overkill.
At least it requires a README to explaint what shouldbe done when a new fix is introduced.
And then any non required complexity should be removed.

**Comment by ABOSTM** [comment]
Refactor BLE library update:
No more need to have patch file in git, it is regenerated
from zephyr version.

File: scripts/ble_README_template.txt (hunk 25 above)
Comment (ABOSTM) [reply scripts/ble_README_template.txt:50]: added

File: lib/ble_zephyr.patch (hunk 26 above)
Comment (ABOSTM) [reply lib/ble_zephyr.patch]: Reworked (file removed)

**Comment by erwango** [comment]
I can see that ble lib update is squashed with stm32wb cube changes. Is that possible to have separate commits? It would help in reviewing the change set produced.

**Comment by ABOSTM** [comment]
> I can see that ble lib update is squashed with stm32wb cube changes. Is that possible to have separate commits? It would help in reviewing the change set produced.

Done

File: scripts/serie_update.py [hunk 28]
```diff
@@ -649,47 +646,43 @@ def apply_zephyr_patch(self):
...
-                commit.write(f"to version {self.version_update}\n")
-            self.os_cmd(
-                ("git", "commit", "-as", "-F", commit_file_path),
-                cwd=self.zephyr_module_serie_path,
+        # Merge & commit
+        # to clean the .rej files, uncomment line: reject()
+        # reject()
+        if lib:
+            logging.info("%s", "commit BLE library update")
+            commit_msg = "BLE library: "
```
Comment (erwango) [inline scripts/serie_update.py:654]:
```suggestion
            commit_msg = "lib/stm32: "
```

**Review by erwango — CHANGES_REQUESTED** [review]
2 minor comments otherwise sounds good.

Not blocking comment:
I wonder if there's an interest to ask about `generic LL HAL headers ` generation.
I  kind of feel this is mandatory anyway.

File: scripts/serie_update.py [hunk 29]
```diff
@@ -777,40 +720,60 @@ def update_stm32_hal_serie(self):
...
+            cwd=self.stm32cube_temp,
+        )
 
         # 3) get the version of cube which is in the zephyr module
         self.current_version = self.get_zephyr_current_version()
         logging.info(
-            "Version "
-            + self.current_version
-            + " is the original version for "
+            f"Version {self.current_version} is the zephyr version for "
```
[the remark this answers is not in the document]
Comment (ABOSTM) [reply scripts/serie_update.py]: Done

File: scripts/serie_update.py (hunk 28 above)
Comment (ABOSTM) [reply scripts/serie_update.py]: Done

**Comment by ABOSTM** [comment]
> Not blocking comment: I wonder if there's an interest to ask about `generic LL HAL headers ` generation. I kind of feel this is mandatory anyway.

From my point of view it was interesting because it was something new integrated in Cube Update.
That said, I am fine to remove it ... but in a future PR

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]
